core/src/main/scala/org/apache/spark/sql/rapids/tool/store/StageModelManager.scala

```diff
@@ -0,0 +1,164 @@
+/*
+ * Copyright (c) 2024, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.rapids.tool.store
+
+import scala.collection.{mutable, Map}
+import scala.collection.immutable.{SortedSet, TreeSet}
+
+import org.apache.spark.internal.Logging
+import org.apache.spark.scheduler.StageInfo
+import org.apache.spark.sql.rapids.tool.annotation.{Calculated, Since, WallClock}
+
+/**
+ * A class to maintain the stage models.
+ */
+@Since("24.02.3")
+class StageModelManager extends Logging {
+  // TODO: 1- Support multithreading and concurrency.
+  //       2- Convert the storage into generic that can be replaced by file-backed storage later
+  //          when needed.
+
+  // A nested HashMap to map between ((Int: stageId, Int: attemptId) -> StageModel).
+  // We keep track of the attemptId to allow improvement down the road if we decide to handle
+  // different Attempts.
+  // - 1st level maps between [Int: StageId -> 2nd Level]
+  // - 2nd level maps between [Int:, StageModel]
+  // Use Nested Maps to store stageModels which should be faster to retrieve than a map of
+  // of composite key (i.e., Tuple).
+  // Composite keys would cost more because it implicitly allocates a new object every time there
+  // is a read operation from the map.
+  private val stageIdToInfo: mutable.HashMap[Int, mutable.HashMap[Int, StageModel]] =
+    new mutable.HashMap[Int, mutable.HashMap[Int, StageModel]]()
+
+  // Holds the mapping between AccumulatorIDs to Stages (1-to-N)
+  // [Long: AccumId -> SortedSet[Int: StageId]]
+  // Note that we keep it as primitive type in case we receive a stageID that does not exist
+  // in the stageIdToInfo map.
+  private val accumIdToStageId: mutable.HashMap[Long, SortedSet[Int]] =
```

Comment:
Just to clarify- This is updated on both TaskEnd and StageCompleted, correct?

Reply:
Yes, 

- For `EventProcessBase.doSparkListenerTaskEnd`

```
    app.stageManager.addAccumIdToStage(
      event.stageId, event.taskInfo.accumulables.map(_.id))
```

- For `EventProcessBase.doSparkListenerStageCompleted`, `getOrCreateStage()` -- as part of fetching the StageModel object --  will loop through the accumulatorIds within the StageInfo and update the accumulator-to-stage-map 

```
    app.getOrCreateStage(event.stageInfo)
```